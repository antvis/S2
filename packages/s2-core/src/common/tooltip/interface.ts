import { SpreadSheet } from '../../sheet-type';
import { SortParam } from '../interface';

export type DataItem = Record<string, any>;

export interface IMenu {
  readonly id: string;
  readonly icon?: any;
  readonly text?: string;
  readonly children?: IMenu[]; // subMenu
}

export interface OperatorProps {
  readonly onClick: (...params) => void;
  readonly menus: IMenu[];
}

export interface Position {
  readonly x: number;
  readonly y: number;
  readonly tipHeight?: number;
}

export type ListItem = {
  name: string;
  value: string | number;
  icon?: string;
};

export interface SortQuery {
  [key: string]: string;
}

export interface TooltipOptions {
  hideSummary?: boolean;
  // button action on the top
  operator?: OperatorProps;
  enterable?: boolean;
  // totals or not
  isTotals?: boolean;
  showSingleTips?: boolean;
  [key: string]: any;
}

export interface SummaryProps {
  readonly name: string;
  readonly value: number | string;
  readonly selectedData: DataItem[];
}

export interface TipsProps {
  readonly tips: string;
}

export interface OperationProps {
  plot: SpreadSheet;
  sortFieldId: string;
  sortQuery: SortQuery;
}

export interface OperationState {
  sortParam: SortParam;
}

export type DetailProps = {
  list: ListItem[];
};

export type InterpretationProps = {
  name: string;
  icon?: any;
  text?: string;
  render?: React.ElementType;
};

export type InfosProps = {
  infos: string;
};

export type ShowProps = {
  position: Position;
  data?: DataProps;
  options?: TooltipOptions;
  element?: React.ReactElement;
};

export type DataProps = {
  summaries?: SummaryProps[];
  details?: ListItem[];
  headInfo?: HeadInfo;
  tips?: string;
  infos?: string;
  interpretation?: InterpretationProps;
};

export type HeadInfo = {
  rows: ListItem[];
  cols: ListItem[];
};
