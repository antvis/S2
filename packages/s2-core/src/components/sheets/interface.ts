import {
  S2DataConfig,
  S2Options,
  CellScrollPosition,
  TargetCellInfo,
  LayoutCol,
  LayoutRow,
  ListSortParams,
  TargetLayoutNode,
  S2Constructor,
} from '@/common/interface';
import { DrillDownProps } from '@/components/drill-down';
import { HeaderCfgProps } from '@/components/header';
import { ThemeCfg } from '@/common/interface';
import { SpreadSheet } from '@/sheet-type';
import { Node } from '@/facet/layout/node';

export type SheetType = 'pivot' | 'table' | 'gridAnalysis';

export interface SpreadsheetProps extends BaseSheetProps {
  sheetType?: SheetType;
}

export interface PartDrillDownInfo {
  // The data of drill down
  drillData: Record<string, string | number>[];
  // The field of drill down
  drillField: string;
}

export interface PartDrillDown {
  // The configuration of drill down
  drillConfig: DrillDownProps;
  // The numbers of drill down result
  drillItemsNum?: number;
  fetchData: (meta: Node, drillFields: string[]) => Promise<PartDrillDownInfo>;
  // Clear the info of drill down
  clearDrillDown?: {
    rowId: string;
  };
  // Decide the drill down icon show conditions.
  displayCondition?: (meta: Node) => boolean;
}

// 用于和下钻组件进行交互联动
export interface PartDrillDownDataCache {
  // 执行下钻的行头id
  rowId: string;
  // 下钻的行头level
  drillLevel: number;
  // 下钻的维度
  drillField: string;
  // 下钻的数据
  drillData: Record<string, string | number>[];
}

export interface PartDrillDownFieldInLevel {
  // 下钻的维度
  drillField: string;
  // 下钻的层级
  drillLevel: number;
}

export interface BaseSheetProps {
  spreadsheet?: (...args: S2Constructor) => SpreadSheet;
  dataCfg: S2DataConfig;
  options: S2Options;
  isLoading?: boolean;
  partDrillDown?: PartDrillDown;
  adaptive?: boolean;
  showPagination?: boolean;
  themeCfg?: ThemeCfg;
  header?: HeaderCfgProps;
  rowLevel?: number;
  colLevel?: number;
  onListSort?: (params: ListSortParams) => void;
  onRowColLayout?: (rows: LayoutRow[], cols: LayoutCol[]) => void;
  onRowCellScroll?: (reachedRow: TargetLayoutNode) => void;
  onColCellScroll?: (reachedCol: TargetLayoutNode) => void;
  onCellScroll?: (position: CellScrollPosition) => void;
  onRowCellClick?: (data: TargetCellInfo) => void;
  onRowCellDoubleClick?: (data: TargetCellInfo) => void;
  onColCellClick?: (data: TargetCellInfo) => void;
  onColCellDoubleClick?: (data: TargetCellInfo) => void;
  onCornerCellClick?: (data: TargetCellInfo) => void;
  onDataCellClick?: (data: TargetCellInfo) => void;
  onDataCellDoubleClick?: (data: TargetCellInfo) => void;
  onDataCellMouseUp?: (data: TargetCellInfo) => void;
  onMergedCellsClick?: (data: TargetCellInfo) => void;
  onMergedCellsDoubleClick?: (data: TargetCellInfo) => void;
  onContextMenu?: (data: TargetCellInfo) => void;
  getSpreadSheet?: (spreadsheet: SpreadSheet) => void;
}
