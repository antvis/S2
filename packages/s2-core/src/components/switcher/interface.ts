import { FieldType } from './constant';

export interface SwitcherItem {
  id: string;
  displayName?: string;
  checked?: boolean;
  children?: Omit<SwitcherItem, 'children'>[];
}

export interface SwitcherState {
  [FieldType.Rows]?: SwitcherItem[];
  [FieldType.Cols]?: SwitcherItem[];
  [FieldType.Values]?: SwitcherItem[];
}

export interface SwitcherResult {
  [FieldType.Rows]: string[];
  [FieldType.Cols]: string[];
  [FieldType.Values]: string[];
  hiddenValues: string[][];
}
