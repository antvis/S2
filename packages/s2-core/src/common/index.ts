export * from './constant';
export * from './debug';
export * from './i18n';
export * from './icons';
export * from './interface';
export * from './store';
