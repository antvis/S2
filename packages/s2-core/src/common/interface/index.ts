export * from './basic';
export * from './condition';
export * from './emitter';
export * from './events';
export * from './frame';
export * from './hooks';
export * from './interaction';
export * from './node';
export * from './resize';
export * from './s2DataConfig';
export * from './s2Options';
export * from './scroll';
export * from './store';
export * from './theme';
export * from './tooltip';
