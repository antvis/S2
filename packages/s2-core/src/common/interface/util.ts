export type DeepRequired<T extends Record<string, any>> = {
  [K in keyof T]-?: NonNullable<T[K]> extends Record<string, any>
    ? DeepRequired<NonNullable<T[K]>>
    : NonNullable<T[K]>;
};
