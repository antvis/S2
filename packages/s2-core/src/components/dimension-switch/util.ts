import { DimensionType, DimensionItem } from './dimension';

export interface OperatedType {
  [type: string]: string[];
}

export const getDimensionsByPredicate = (
  data: DimensionType[],
  predicate: (value: DimensionItem) => boolean,
) => {
  return data.reduce((obj, t) => {
    obj[t.type] = t.items.filter(predicate).map((i) => i.id);
    return obj;
  }, {} as OperatedType);
};
