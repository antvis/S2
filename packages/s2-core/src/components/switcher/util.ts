import { filter, flatten, isEmpty, isNil, map } from 'lodash';
import { DraggableLocation } from 'react-beautiful-dnd';
import {
  FieldType,
  MAX_DIMENSION_COUNT,
  SWITCHER_PREFIX_CLS,
} from './constant';
import { SwitcherItem, SwitchResult, SwitchState } from './interface';
import { getClassNameWithPrefix } from '@/utils/get-classnames';

export const getSwitcherClassName = (...classNames: string[]) =>
  getClassNameWithPrefix(SWITCHER_PREFIX_CLS, ...classNames);

export const getNonEmptyFieldCount = (state: SwitchState) => {
  return [
    state[FieldType.Rows],
    state[FieldType.Cols],
    state[FieldType.Values],
  ].reduce((sum, value) => sum + (isEmpty(value) ? 0 : 1), 0);
};

export const getMainLayoutClassName = (nonEmptyCount: number) => {
  switch (nonEmptyCount) {
    case 1:
      return getSwitcherClassName('content', 'one-dimension');
    case 2:
      return getSwitcherClassName('content', 'two-dimensions');
    default:
      return getSwitcherClassName('content', 'three-dimensions');
  }
};

export const showDimensionCrossRows = (nonEmptyCount: number) =>
  nonEmptyCount < MAX_DIMENSION_COUNT;

export const isMeasureType = (fieldType: FieldType) =>
  fieldType === FieldType.Values;

export const moveItem = (
  source: SwitcherItem[],
  destination: SwitcherItem[],
  droppableSource: DraggableLocation,
  droppableDestination: DraggableLocation,
): SwitchState => {
  // change order in same column
  if (droppableDestination.droppableId === droppableSource.droppableId) {
    const updatingDestination = [...destination];
    const [removed] = updatingDestination.splice(droppableSource.index, 1);
    updatingDestination.splice(droppableDestination.index, 0, removed);
    return {
      [droppableDestination.droppableId]: updatingDestination,
    };
  }
  // move to other column
  const updatingSource = [...source];
  const updatingDestination = [...destination];

  const [removed] = updatingSource.splice(droppableSource.index, 1);
  updatingDestination.splice(droppableDestination.index, 0, removed);

  return {
    [droppableSource.droppableId]: updatingSource,
    [droppableDestination.droppableId]: updatingDestination,
  };
};

export const checkItem = (
  source: SwitcherItem[],
  checked: boolean,
  id: string,
  parentId?: string,
): SwitcherItem[] => {
  const target: SwitcherItem = {
    ...source.find((item) => item.id === (parentId ?? id)),
  };

  // 有 parentId 时，说明是第二层次项的改变
  if (parentId) {
    target.children = map(target.children, (item) => ({
      ...item,
      checked: item.id === id ? checked : item.checked,
    }));
  } else {
    target.checked = checked;
    target.children = map(target.children, (item) => ({
      ...item,
      checked: checked,
    }));
  }

  return source.map((item) => (item.id === target.id ? target : item));
};

export const generateSwitchResult = (state: SwitchState): SwitchResult => {
  // rows and cols can't be hidden
  const rows = map(state[FieldType.Rows], 'id');
  const cols = map(state[FieldType.Cols], 'id');

  // flatten all values and derived values
  const values = flatten(
    map(state[FieldType.Values], (item) => {
      const children = map(item.children, 'id');
      return [item.id, ...children];
    }),
  );

  const filterHiddenValues = (item: SwitcherItem) => item.checked === false;

  //  get all hidden values
  const hiddenValues = flatten(
    map(filter(state[FieldType.Values], filterHiddenValues), (item) => {
      const hiddenChildren = map(
        filter(item.children, filterHiddenValues),
        'id',
      );
      return [item.id, ...hiddenChildren];
    }),
  );

  return { rows, cols, values, hiddenValues };
};
