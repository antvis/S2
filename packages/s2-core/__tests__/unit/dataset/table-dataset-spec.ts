/**
 * table mode data-set test.
 */
import { orderBy, uniq } from 'lodash';
import { assembleDataCfg } from '../../util';
import { S2DataConfig } from '@/common/interface';
import { TableSheet } from '@/sheet-type';
import { TableDataSet } from '@/data-set/table-data-set';

jest.mock('src/sheet-type');
jest.mock('src/facet/layout/node');
const MockTableSheet = TableSheet as any as jest.Mock<TableSheet>;

describe('Table Mode Dataset Test', () => {
  let dataSet: TableDataSet;
  const dataCfg: S2DataConfig = {
    ...assembleDataCfg({}),
    meta: [],
    fields: {
      columns: ['province', 'city', 'type', 'sub_type', 'number'],
    },
  };
  beforeEach(() => {
    MockTableSheet.mockClear();
    dataSet = new TableDataSet(new MockTableSheet());

    dataSet.setDataCfg(dataCfg);
  });

  describe('test base dataset structure', () => {
    test('should get correct field data', () => {
      expect(dataSet.fields.rows).toEqual(undefined);
      expect(dataSet.fields.columns).toEqual([
        'province',
        'city',
        'type',
        'sub_type',
        'number',
      ]);
      expect(dataSet.fields.values).toEqual(undefined);
    });

    test('should get correct meta data', () => {
      expect(dataSet.meta).toEqual(expect.objectContaining([]));
    });
  });

  describe('test for query data', () => {
    test('getCellData function', () => {
      expect(
        dataSet.getCellData({
          query: {
            rowIndex: 0,
            col: 'city',
          },
        }),
      ).toEqual('杭州市');

      expect(
        dataSet.getCellData({
          query: {
            rowIndex: 2,
            col: 'number',
          },
        }),
      ).toEqual(3877);

      expect(
        dataSet.getCellData({
          query: {
            rowIndex: 5,
            col: 'sub_type',
          },
        }),
      ).toEqual('沙发');
    });
  });

  describe('test dataset query when no data', () => {
    test('should handle getCellData when no data', () => {
      MockTableSheet.mockClear();
      const emptyDataSet = new TableDataSet(new MockTableSheet());
      emptyDataSet.setDataCfg({
        ...dataCfg,
        data: [],
      });
      expect(
        emptyDataSet.getCellData({
          query: {
            col: 'sub_type',
            rowIndex: 0,
          },
        }),
      ).toEqual(undefined);
    });
  });

  describe('test for sort and filter', () => {
    it('should getCellData with filteredValues', () => {
      dataSet.setDataCfg({
        ...dataCfg,
        filterParams: [
          {
            filterKey: 'province',
            filteredValues: ['浙江省'],
          },
        ],
      });
      expect(
        dataSet.getCellData({
          query: {
            rowIndex: 0,
            col: 'city',
          },
        }),
      ).toEqual('成都市');
    });
    it('should getCellData with customFilter', () => {
      dataSet.setDataCfg({
        ...dataCfg,
        filterParams: [
          {
            filterKey: 'province',
            customFilter: (row) => row.province === '浙江省',
          },
        ],
      });
      expect(
        dataSet.getCellData({
          query: {
            rowIndex: 0,
            col: 'city',
          },
        }),
      ).toEqual('杭州市');
    });

    it('should getCellData with customFilter and filteredValues', () => {
      dataSet.setDataCfg({
        ...dataCfg,
        filterParams: [
          {
            filterKey: 'province',
            filteredValues: ['浙江省'],
            customFilter: (row) =>
              row.province === '浙江省' || row.province === '四川省',
          },
        ],
      });
      expect(
        dataSet.getCellData({
          query: {
            rowIndex: 0,
            col: 'city',
          },
        }),
      ).toEqual('成都市');
    });

    it('should getCellData with sort', () => {
      dataSet.setDataCfg({
        ...dataCfg,
        sortParams: [
          {
            sortFieldId: 'number',
            sortMethod: 'ASC',
          },
        ],
      });
      expect(
        dataSet.getCellData({
          query: {
            rowIndex: 0,
            col: 'number',
          },
        }),
      ).toEqual(245);
    });

    it('should get correct sorted data with sortFunc', () => {
      const sortFieldId = 'number';
      const sortMethod = 'desc';
      const sorted = orderBy(
        dataSet.getDisplayDataSet(),
        [sortFieldId],
        [sortMethod.toLocaleLowerCase() as 'asc' | 'desc'],
      );
      dataSet.sortParams = [
        {
          sortFieldId,
          sortMethod,
          sortFunc: ({ data, sortMethod, sortFieldId }) => {
            return orderBy(
              data,
              [sortFieldId],
              [sortMethod.toLocaleLowerCase() as 'asc' | 'desc'],
            );
          },
        },
      ];
      dataSet.handleDimensionValuesSort();

      expect(dataSet.getDisplayDataSet()).toStrictEqual(sorted);
    });

    it('should get correct sorted data with sortBy', () => {
      const sortFieldId = 'city';
      const sortBy = [
        '杭州市',
        '成都市',
        '绍兴市',
        '舟山市',
        '宁波市',
        '绵阳市',
        '乐山市',
        '南充市',
      ];
      dataSet.sortParams = [
        {
          sortFieldId,
          sortBy,
          sortFunc: null,
        },
      ];
      dataSet.handleDimensionValuesSort();
      expect(
        uniq(dataSet.getDisplayDataSet().map((item) => item.city)),
      ).toStrictEqual(sortBy);
    });

    it('should get correct sorted data with sortMethod', () => {
      const sortFieldId = 'number';

      let result = orderBy(dataSet.getDisplayDataSet(), [sortFieldId], 'asc');
      dataSet.sortParams = [
        {
          sortFieldId,
          sortMethod: 'asc',
        },
      ];
      dataSet.handleDimensionValuesSort();
      expect(result).toStrictEqual(dataSet.getDisplayDataSet());

      result = orderBy(dataSet.getDisplayDataSet(), [sortFieldId], 'desc');
      dataSet.sortParams = [
        {
          sortFieldId,
          sortMethod: 'desc',
        },
      ];
      dataSet.handleDimensionValuesSort();
      expect(result).toStrictEqual(dataSet.getDisplayDataSet());
    });

    it('should get scoped sort result', () => {
      const sortFieldId = 'number';
      const query = {
        city: '杭州市',
      };

      const rest = dataSet
        .getDisplayDataSet()
        .filter((record) => record.city !== '杭州市');

      const result = orderBy(
        dataSet
          .getDisplayDataSet()
          .filter((record) => record.city === '杭州市'),
        [sortFieldId],
        'asc',
      );

      dataSet.sortParams = [
        {
          sortFieldId,
          sortMethod: 'asc',
          query,
        },
      ];
      dataSet.handleDimensionValuesSort();
      expect([...result, ...rest]).toStrictEqual(dataSet.getDisplayDataSet());
    });
  });
});
