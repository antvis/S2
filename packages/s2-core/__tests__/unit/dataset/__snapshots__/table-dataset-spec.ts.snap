// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Mode Dataset Test test for sort and filter should asc sort by number field 1`] = `
Array [
  Object {
    "city": "绵阳市",
    "number": 245,
    "province": "四川省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "乐山市",
    "number": 352,
    "province": "四川省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "绍兴市",
    "number": 632,
    "province": "浙江省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "舟山市",
    "number": 834,
    "province": "浙江省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "杭州市",
    "number": 945,
    "province": "浙江省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "宁波市",
    "number": 1145,
    "province": "浙江省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "绍兴市",
    "number": 1304,
    "province": "浙江省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "杭州市",
    "number": 1343,
    "province": "浙江省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "绍兴市",
    "number": 1354,
    "province": "浙江省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "舟山市",
    "number": 1432,
    "province": "浙江省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "宁波市",
    "number": 1523,
    "province": "浙江省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "舟山市",
    "number": 1634,
    "province": "浙江省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "成都市",
    "number": 1723,
    "province": "四川省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "绵阳市",
    "number": 1822,
    "province": "四川省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "南充市",
    "number": 1943,
    "province": "四川省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "绵阳市",
    "number": 2244,
    "province": "四川省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "乐山市",
    "number": 2330,
    "province": "四川省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "南充市",
    "number": 2333,
    "province": "四川省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "成都市",
    "number": 2335,
    "province": "四川省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "绍兴市",
    "number": 2367,
    "province": "浙江省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "乐山市",
    "number": 2445,
    "province": "四川省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "成都市",
    "number": 2451,
    "province": "四川省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "南充市",
    "number": 2457,
    "province": "四川省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "乐山市",
    "number": 2458,
    "province": "四川省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "绵阳市",
    "number": 3077,
    "province": "四川省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "南充市",
    "number": 3551,
    "province": "四川省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "宁波市",
    "number": 3877,
    "province": "浙江省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "成都市",
    "number": 4004,
    "province": "四川省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "舟山市",
    "number": 4342,
    "province": "浙江省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "杭州市",
    "number": 5343,
    "province": "浙江省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "宁波市",
    "number": 7234,
    "province": "浙江省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "杭州市",
    "number": 7789,
    "province": "浙江省",
    "sub_type": "桌子",
    "type": "家具",
  },
]
`;

exports[`Table Mode Dataset Test test for sort and filter should desc sort by number field 1`] = `
Array [
  Object {
    "city": "杭州市",
    "number": 7789,
    "province": "浙江省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "宁波市",
    "number": 7234,
    "province": "浙江省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "杭州市",
    "number": 5343,
    "province": "浙江省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "舟山市",
    "number": 4342,
    "province": "浙江省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "成都市",
    "number": 4004,
    "province": "四川省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "宁波市",
    "number": 3877,
    "province": "浙江省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "南充市",
    "number": 3551,
    "province": "四川省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "绵阳市",
    "number": 3077,
    "province": "四川省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "乐山市",
    "number": 2458,
    "province": "四川省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "南充市",
    "number": 2457,
    "province": "四川省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "成都市",
    "number": 2451,
    "province": "四川省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "乐山市",
    "number": 2445,
    "province": "四川省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "绍兴市",
    "number": 2367,
    "province": "浙江省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "成都市",
    "number": 2335,
    "province": "四川省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "南充市",
    "number": 2333,
    "province": "四川省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "乐山市",
    "number": 2330,
    "province": "四川省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "绵阳市",
    "number": 2244,
    "province": "四川省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "南充市",
    "number": 1943,
    "province": "四川省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "绵阳市",
    "number": 1822,
    "province": "四川省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "成都市",
    "number": 1723,
    "province": "四川省",
    "sub_type": "桌子",
    "type": "家具",
  },
  Object {
    "city": "舟山市",
    "number": 1634,
    "province": "浙江省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "宁波市",
    "number": 1523,
    "province": "浙江省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "舟山市",
    "number": 1432,
    "province": "浙江省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "绍兴市",
    "number": 1354,
    "province": "浙江省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "杭州市",
    "number": 1343,
    "province": "浙江省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "绍兴市",
    "number": 1304,
    "province": "浙江省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "宁波市",
    "number": 1145,
    "province": "浙江省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "杭州市",
    "number": 945,
    "province": "浙江省",
    "sub_type": "笔",
    "type": "办公用品",
  },
  Object {
    "city": "舟山市",
    "number": 834,
    "province": "浙江省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "绍兴市",
    "number": 632,
    "province": "浙江省",
    "sub_type": "沙发",
    "type": "家具",
  },
  Object {
    "city": "乐山市",
    "number": 352,
    "province": "四川省",
    "sub_type": "纸张",
    "type": "办公用品",
  },
  Object {
    "city": "绵阳市",
    "number": 245,
    "province": "四川省",
    "sub_type": "笔",
    "type": "办公用品",
  },
]
`;
